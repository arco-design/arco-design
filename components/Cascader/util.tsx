import { isArray } from '../_util/is';
import Store from './base/store';
import { CascaderProps } from './interface';

export const ValueSeparator = '__arco_cascader__';

export const SHOW_PARENT = 'parent';
export const SHOW_CHILD = 'child';

export const PANEL_MODE = {
  cascader: 'cascader',
  select: 'select',
};

export function isEmptyValue(value) {
  return !value || (isArray(value) && value.length === 0);
}

export function getConfig(props: CascaderProps) {
  return {
    showEmptyChildren: props.showEmptyChildren,
    changeOnSelect: props.changeOnSelect,
    lazyload: !!props.loadMore,
    fieldNames: props.fieldNames,
    filterOption: props.filterOption,
    showParent:
      props.mode === 'multiple' && !props.changeOnSelect && props.checkedStrategy === SHOW_PARENT,
  };
}

export function getStore(props, value) {
  const tmp = value ? (Array.isArray(value[0]) ? value : [value]) : [];
  return new Store(props.options || [], tmp, getConfig(props));
}

export const transformValuesToSet = (values: string[][]) => {
  const _values = values || [];
  const valuesSet = _values.reduce((set, next) => {
    // 'next' could be a string.
    set.add([].concat(next).join(ValueSeparator));
    return set;
  }, new Set());

  return valuesSet;
};

export const valueInSet = (set, value: string[]) => {
  const _value = value || [];
  return set.has(_value.join(ValueSeparator));
};

export const removeValueFromSet = (set, value: string[]) => {
  const _value = value || [];
  return set.delete(_value.join(ValueSeparator));
};

export const formatValue = (value, isMultiple, store?): string[][] | undefined => {
  let _value = [];
  if (value === undefined) {
    _value = [];
  } else if (isMultiple) {
    _value = value;
  } else {
    _value = [value];
  }

  if (store && store.config.showParent) {
    const checkedNodes = store.getCheckedNodes();
    const valuesSet = transformValuesToSet(checkedNodes.map((node) => node.pathValue));
    const result = [];
    const temp = {};
    _value.map((v) => {
      v.some((_, index, arr) => {
        const curVal = arr.slice(0, index + 1);
        const pass = valueInSet(valuesSet, curVal);
        if (pass && !temp[curVal.join(ValueSeparator)]) {
          result.push(curVal);
          temp[curVal.join(ValueSeparator)] = 1;
        }
        return pass;
      });
    });

    return result;
  }
  return _value;
};

// change check status to false
const deny2Checked = (option) => {
  const deny = (options) => {
    return !Array.isArray(options)
      ? false
      : options.every((item) => {
          if (item._checked || item.disabled) {
            return true;
          }
          return deny(item.children);
        });
  };
  return option._halfChecked && deny(option?.children);
};

export const getMultipleCheckValue = (propsValue, store: Store<any>, option, _checked) => {
  const checked = _checked && deny2Checked(option) ? false : _checked;

  const beforeValueSet = store.getCheckedNodes().reduce((set, node) => {
    set.add(node.pathValue.join(ValueSeparator));
    return set;
  }, new Set());

  option.setCheckedState(checked);
  const checkedNodes = store.getCheckedNodes();
  const currentValue = checkedNodes.map((node) => node.pathValue);
  const currentValueSet = transformValuesToSet(currentValue);

  const newValueSet = new Set();
  return propsValue
    .filter((v) => {
      // v 不在 beforeValueSet 中，说明 v 不包含对应的option。直接返回true，不应该清除掉。
      if (!valueInSet(beforeValueSet, v) || valueInSet(currentValueSet, v)) {
        newValueSet.add(v.join(ValueSeparator));
        return true;
      }
    })
    .concat(
      currentValue.filter((v) => {
        return !valueInSet(newValueSet, v);
      })
    );
};
