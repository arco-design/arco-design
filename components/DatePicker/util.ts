import dayjs, { Dayjs } from 'dayjs';
import { isArray, isDayjs } from '../_util/is';
import { methods } from '../_util/dayjs';

function getFormat(time) {
  return isDayjs(time) && time.format('HH:mm:ss');
}

export function isTimeArrayChange(prevTime: Dayjs[], nextTime: Dayjs[]) {
  return (
    getFormat(prevTime[0]) !== getFormat(nextTime[0]) ||
    getFormat(prevTime[1]) !== getFormat(nextTime[1])
  );
}

export function getAvailableDayjsLength(value) {
  if (!value) {
    return 0;
  }
  if (isArray(value)) {
    if (isDayjs(value[0]) && isDayjs(value[1])) {
      return 2;
    }
    if (!isDayjs(value[0]) && !isDayjs(value[1])) {
      return 0;
    }
    return 1;
  }
  return 0;
}

// https://github.com/react-component/picker/blob/master/src/utils/dateUtil.ts#L234
export function isDisabledDate(cellDate, disabledDate, mode): boolean {
  if (typeof disabledDate !== 'function') {
    return false;
  }
  // Whether cellDate is disabled in range
  const getDisabledFromRange = (
    currentMode: 'date' | 'month' | 'year',
    start: number,
    end: number
  ) => {
    let current = start;
    while (current <= end) {
      let date: Dayjs;
      switch (currentMode) {
        case 'date': {
          date = methods.set(cellDate, 'date', current);
          if (!disabledDate(date)) {
            return false;
          }
          break;
        }
        case 'month': {
          date = methods.set(cellDate, 'month', current);
          if (!isDisabledDate(date, disabledDate, 'month')) {
            return false;
          }
          break;
        }
        case 'year': {
          date = methods.set(cellDate, 'year', current);
          if (!isDisabledDate(date, disabledDate, 'year')) {
            return false;
          }
          break;
        }
        default:
          break;
      }
      current += 1;
    }
    return true;
  };
  switch (mode) {
    case 'date':
    case 'week': {
      return disabledDate(cellDate);
    }
    case 'month': {
      const startDate = 1;
      const endDate = cellDate.endOf('month').get('date');
      return getDisabledFromRange('date', startDate, endDate);
    }
    case 'quarter': {
      const startMonth = Math.floor(cellDate.get('month') / 3) * 3;
      const endMonth = startMonth + 2;
      return getDisabledFromRange('month', startMonth, endMonth);
    }
    case 'year': {
      return getDisabledFromRange('month', 0, 11);
    }
    default:
      return false;
  }
}

type WeekStartType = 0 | 1 | 2 | 3 | 4 | 5 | 6;

export function getDefaultWeekStart(dayjsLocale: string): WeekStartType {
  return (dayjs.Ls?.[dayjsLocale]?.weekStart as WeekStartType) || 0;
}

export function getLocaleDayjsValue(
  date: Dayjs | undefined,
  dayjsLocale: string
): Dayjs | undefined {
  return date ? date.locale(dayjsLocale) : date;
}
