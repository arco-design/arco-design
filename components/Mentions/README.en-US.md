`````
Component / Data Entry

# Mentions

Mentions is used to mention someone or something in the input.
`````

%%Content%%

## API

### Mentions

|Property|Description|Type|DefaultValue|Version|
|---|---|---|---|---|
|alignTextarea|Whether the popup is aligned with the input|boolean |`true`|-|
|allowClear|Whether allow clear the content|boolean |`-`|2.2.0|
|disabled|Whether the textarea is disabled|boolean |`-`|-|
|error|Whether the textarea is error.(Deprecated, removed in the next major version, use status='error' instead)|boolean |`-`|-|
|defaultValue|To set default value|string |`-`|-|
|placeholder|textarea placeholder|string |`-`|-|
|split|Set split string before and after selected mention|string |`-`|-|
|value|To set value|string |`-`|-|
|position|Position of dropdown|'top' \| 'tl' \| 'tr' \| 'bottom' \| 'bl' \| 'br' |`bl`|-|
|status|Status|'error' \| 'warning' |`-`|2.45.0|
|notFoundContent|The content displayed when there is no data|ReactNode |`-`|-|
|autoSize|Height autoSize feature|boolean \| { minRows?: number; maxRows?: number } |`-`|-|
|className|Additional css class|string \| string[] |`-`|-|
|options|Options of dropdown|(\| string\| number\| { label: ReactNode \| string; value: string \| number; disabled?: boolean })[] |`-`|-|
|prefix|Set trigger prefix keyword|string \| string[] |``@``|-|
|style|Additional style|CSSProperties |`-`|-|
|triggerProps|All `Trigger` component props|Partial&lt;[TriggerProps](trigger#trigger)&gt; |`-`|-|
|wrapperStyle|With `showWordLimit`, a `div` will be outside the `textarea` tag, and `wrapperStyle` is used to configure the style of it.|CSSProperties |`-`|-|
|filterOption|Customize filter option logic|false \| ((inputValue: string, option) => boolean) |`-`|-|
|getPopupContainer|Set the mount HTML node for suggestions|(node: HTMLElement) => HTMLElement |`-`|-|
|onBlur|Trigger when mentions lose focus|(e) => void |`-`|-|
|onChange|Callback when input value is changed|(value: string) => void |`-`|-|
|onClear|Callback when click clear button|() => void |`-`|2.2.0|
|onFocus|Trigger when mentions get focus|(e) => void |`-`|-|
|onPressEnter|Callback when press enter key|(e) => void |`-`|-|
|onSearch|Callback on search|(text: string, prefix: string) => void |`-`|-|

