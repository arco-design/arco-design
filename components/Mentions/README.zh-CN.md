`````
组件 / 数据输入

# 提及 Mentions

用于在输入中提及某人或某事，常用于发布、聊天或评论功能。
`````

%%Content%%

## API

### Mentions

|参数名|描述|类型|默认值|版本|
|---|---|---|---|---|
|alignTextarea|弹出框是否与输入框对齐|boolean |`true`|-|
|allowClear|允许清空输入框|boolean |`-`|2.2.0|
|disabled|是否禁用|boolean |`-`|-|
|error|是否是错误状态。(废弃，下个大版本移除，使用 status='error' 替代)|boolean |`-`|-|
|defaultValue|输入框默认值|string |`-`|-|
|placeholder|输入框提示文字|string |`-`|-|
|split|选中项前后分隔符|string |`-`|-|
|value|输入框的值|string |`-`|-|
|position|下拉框的弹出位置|'top' \| 'tl' \| 'tr' \| 'bottom' \| 'bl' \| 'br' |`bl`|-|
|status|状态|'error' \| 'warning' |`-`|2.45.0|
|notFoundContent|下拉列表没有数据时显示的内容|ReactNode |`-`|-|
|autoSize|是否自动调整输入框的高度|boolean \| { minRows?: number; maxRows?: number } |`-`|-|
|className|节点类名|string \| string[] |`-`|-|
|options|下拉框可选项|(\| string\| number\| { label: ReactNode \| string; value: string \| number; disabled?: boolean })[] |`-`|-|
|prefix|触发关键字|string \| string[] |``@``|-|
|style|节点样式|CSSProperties |`-`|-|
|triggerProps|可以接受所有 Trigger 组件的 Props|Partial&lt;[TriggerProps](trigger#trigger)&gt; |`-`|-|
|wrapperStyle|开启字数统计之后，会在 `textarea` 标签外包一层 `div`，`wrapperStyle` 用来配置这个 `div` 的样式。|CSSProperties |`-`|-|
|filterOption|是否根据输入的值筛选数据，可传入函数自定义过滤逻辑。|false \| ((inputValue: string, option) => boolean) |`-`|-|
|getPopupContainer|弹出框挂载的父节点|(node: HTMLElement) => HTMLElement |`-`|-|
|onBlur|失焦时的回调|(e) => void |`-`|-|
|onChange|输入改变时的回调|(value: string) => void |`-`|-|
|onClear|点击清除按钮的回调|() => void |`-`|2.2.0|
|onFocus|聚焦时的回调|(e) => void |`-`|-|
|onPressEnter|按下回车键的回调|(e) => void |`-`|-|
|onSearch|搜索时的回调|(text: string, prefix: string) => void |`-`|-|

