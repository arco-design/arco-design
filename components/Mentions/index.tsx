import React, { useContext, useEffect, useRef, useState } from 'react';
import cs from '../_util/classNames';
import { ConfigContext } from '../ConfigProvider';
import Input from '../Input';
import Select from '../Select';
import { MentionsProps } from './interface';
import { ArrowDown, ArrowUp, Enter, Esc } from '../_util/keycode';
import { getBeforeSelectionText, getLastMeasureIndex, isValidSearch } from './utils';
import useMergeValue from '../_util/hooks/useMergeValue';
import useMergeProps from '../_util/hooks/useMergeProps';

const TextArea = Input.TextArea;

const FunctionalKeyCodeList = [Esc.code, Enter.code, ArrowUp.code, ArrowDown.code];

const defaultProps: MentionsProps = {
  prefix: '@',
  split: ' ',
  rows: 1,
  position: 'bl',
  alignTextarea: true,
};

function Mentions(baseProps: MentionsProps, ref) {
  const { getPrefixCls, componentConfig } = useContext(ConfigContext);
  const props = useMergeProps<MentionsProps>(baseProps, defaultProps, componentConfig?.Mentions);
  const {
    style,
    className,
    options,
    prefix,
    split,
    position,
    alignTextarea,
    filterOption,
    triggerProps,
    getPopupContainer,
    onChange,
    onSearch,
    notFoundContent,
    ...rest
  } = props;

  const prefixCls = getPrefixCls('mentions');

  const refSelect = useRef(null);
  const refMeasure = useRef(null);
  const refTextarea = useRef(null);

  const [value, setValue] = useMergeValue('', {
    value: props.value,
    defaultValue: props.defaultValue,
  });
  const [measureInfo, setMeasureInfo] = useState({
    measuring: false,
    location: 0,
    text: '',
    prefix: '',
  });

  useEffect(() => {
    if (refTextarea.current && refMeasure.current) {
      refMeasure.current.scrollTop = refTextarea.current.dom.scrollTop;
    }
  });

  const stopMeasure = () => {
    setMeasureInfo({
      ...measureInfo,
      measuring: false,
      location: 0,
      text: '',
    });
  };

  const handleOptionSelect = (optionValue) => {
    const measureStart = measureInfo.location;
    const measureEnd = measureInfo.location + measureInfo.text.length;
    let head = value.slice(0, measureStart);
    let tail = value.slice(measureEnd + 1);
    head += !head || head.endsWith(split) || head.endsWith('\n') ? '' : split;
    tail = (!tail || tail.startsWith(split) || tail.startsWith('\n') ? '' : split) + tail;
    // If the content already exists before or after the matched content, add a split character
    const match = `${measureInfo.prefix}${optionValue}`;
    const nextValue = `${head}${match}${tail}`;

    setValue(nextValue);
    stopMeasure();
    onChange && onChange(nextValue);
  };

  const textAreaEventHandlers = {
    onChange: (value) => {
      setValue(value);
      onChange && onChange(value);
    },
    onKeyDown: (event) => {
      const keyCode = event.keyCode || event.which;

      if (keyCode === Esc.code) {
        refTextarea.current && refTextarea.current.blur();
      }

      if (refSelect.current) {
        refSelect.current.hotkeyHandler(event);

        if (keyCode === Enter.code || keyCode === ArrowUp.code || keyCode === ArrowDown.code) {
          event.preventDefault();
        }
      }
    },
    onKeyUp: (event) => {
      const { key, which: keyCode, target } = event;

      // return immediately when hit any one of the function keys
      if (~FunctionalKeyCodeList.indexOf(keyCode)) {
        return;
      }

      const textBeforeSelection = getBeforeSelectionText(target);
      const { location: measureIndex, prefix: measurePrefix } = getLastMeasureIndex(
        textBeforeSelection,
        prefix
      );
      const measureText = textBeforeSelection.slice(measureIndex + measurePrefix.length);

      if (measureIndex > -1 && isValidSearch(measureText, props)) {
        if (key === measurePrefix || measureInfo.measuring || measureText !== measureInfo.text) {
          setMeasureInfo({
            measuring: true,
            text: measureText,
            prefix: measurePrefix,
            location: measureIndex,
          });
        }

        onSearch && onSearch(measureText, measurePrefix);
      } else if (measureInfo.measuring) {
        stopMeasure();
      }
    },
    onBlur: stopMeasure,
  };

  return (
    <div
      ref={ref}
      style={style}
      className={cs(`${prefixCls}`, { [`${prefixCls}-align-textarea`]: alignTextarea }, className)}
    >
      <TextArea
        ref={refTextarea}
        className={`${prefixCls}-textarea`}
        value={value}
        {...textAreaEventHandlers}
        {...rest}
      />
      {measureInfo.measuring && (
        <div ref={refMeasure} className={`${prefixCls}-measure`}>
          {value.slice(0, measureInfo.location)}
          <Select
            ref={refSelect}
            options={options}
            inputValue={measureInfo.text}
            notFoundContent={notFoundContent}
            triggerElement={
              <span className={`${prefixCls}-measure-trigger`}>{measureInfo.prefix}</span>
            }
            triggerProps={{
              popupVisible: true,
              autoAlignPopupWidth: alignTextarea,
              position,
              ...triggerProps,
            }}
            filterOption={filterOption}
            getPopupContainer={getPopupContainer}
            onChange={handleOptionSelect}
          />
          {value.slice(measureInfo.location + measureInfo.prefix.length)}
        </div>
      )}
    </div>
  );
}

const MentionsComponent = React.forwardRef<unknown, MentionsProps>(Mentions);

MentionsComponent.displayName = 'Mentions';

export default MentionsComponent;

export { MentionsProps };
