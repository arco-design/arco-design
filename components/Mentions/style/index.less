@import './token.less';
@mentions-prefix-cls: ~'@{prefix}-mentions';

.@{mentions-prefix-cls} {
  display: inline-block;
  position: relative;
  box-sizing: border-box;
  width: 100%;
  vertical-align: middle;
  white-space: pre-wrap;

  &-textarea,
  &-measure {
    padding: @mentions-padding-vertical @mentions-padding-horizontal;
    font-size: @mentions-font-size;
    line-height: @mentions-line-height;
    white-space: inherit;
  }

  &-textarea {
    resize: none;
  }

  &-measure,
  &-align-textarea &-measure-trigger {
    position: absolute;
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;
    overflow: auto;
    visibility: hidden;
    pointer-events: none;
  }
}

.@{mentions-prefix-cls}-rtl {
  direction: rtl;
}
