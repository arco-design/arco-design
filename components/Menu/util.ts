import React, { ReactElement } from 'react';

export type MenuInfo = {
  keyPath: string[];
  prev: string;
  next: string;
  firstChild?: string;
  lastChild?: string;
  disabled?: boolean;
};

export const PROPS_NEED_TO_BE_PASSED_IN_SUBMENU = ['popup', 'triggerProps', 'selectable'];

// Expand MenuGroup to get an array only contains MenuItem and SubMenu
const flatMenuGroup = (children): ReactElement[] => {
  let validMenuItems = [];

  React.Children.forEach(children, (item) => {
    const menuType = item?.type?.menuType;
    if (menuType === 'MenuItem' || menuType === 'SubMenu') {
      validMenuItems.push(item);
    } else if (menuType === 'MenuGroup') {
      validMenuItems = validMenuItems.concat(flatMenuGroup(item.props.children));
    }
  });

  return validMenuItems;
};

export const generateInfoMap = (
  children,
  keyPath: string[] = [],
  result: {
    [key: string]: MenuInfo;
  } = {}
) => {
  const validChildrenList: any[] = flatMenuGroup(children);

  validChildrenList.forEach((item, index) => {
    const key = item.key;
    const menuType = item.type.menuType;
    const _keyPath = [key, ...keyPath];
    const info: MenuInfo = {
      keyPath: [],
      prev: validChildrenList[index - 1]?.key || null,
      next: validChildrenList[index + 1]?.key || null,
    };

    if (index === 0 || index === validChildrenList.length - 1) {
      const parentKey = _keyPath[1];
      const propertyName = index === 0 ? 'firstChild' : 'lastChild';
      if (parentKey) {
        result[parentKey] = {
          ...result[parentKey],
          [propertyName]: key,
        };
      }
    }

    switch (menuType) {
      case 'SubMenu':
        info.keyPath = _keyPath;
        generateInfoMap(item.props.children, _keyPath, result);
        break;

      case 'MenuItem':
        info.keyPath = _keyPath;
        info.disabled = item.props.disabled;
        break;

      default:
        break;
    }

    result[key] = {
      ...result[key],
      ...info,
    };
  });

  return result;
};

export const processChildren = (
  children,
  props
): Array<Exclude<unknown, boolean | null | undefined>> => {
  return React.Children.map(children, (item, index) => {
    if (!item || !item.props) {
      return item;
    }

    const isHTMLElement = typeof item.type === 'string';
    const isMenuSubComponent = item.type && item.type.menuType;

    // 处理 <div> 包裹 MenuItem 之类的情况
    if (!isMenuSubComponent && item.props.children) {
      const _props = isHTMLElement ? {} : props;
      return React.cloneElement(item, {
        ..._props,
        _key: item.key,
        children: processChildren(item.props.children, props),
      });
    }

    return isHTMLElement
      ? item
      : React.cloneElement(item, {
          ...props,
          // Properties of the component itself have higher priority
          ...item.props,
          _key: item.key || `$menu-${index}`,
        });
  });
};

export function isChildrenSelected(children, keys: string[]) {
  let find = false;
  function loop(_children) {
    if (!_children || find) {
      return;
    }
    React.Children.forEach(_children, (c) => {
      if (c && c.props && c.type && !find) {
        const menuType = c.type.menuType;
        const selectable = c.props.selectable;
        if (menuType === 'MenuItem') {
          find = keys.indexOf(c.key) !== -1;
        } else if (menuType === 'SubMenu' && selectable) {
          find = keys.indexOf(c.key) !== -1;
        }
        if (!find && c.props.children) {
          loop(c.props.children);
        }
      }
    });
  }
  loop(children);
  return find;
}
