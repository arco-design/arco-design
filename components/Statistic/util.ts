import { padStart } from '../_util/pad';

const units: [string, number][] = [
  ['Y', 1000 * 60 * 60 * 24 * 365], // years
  ['M', 1000 * 60 * 60 * 24 * 30], // months
  ['D', 1000 * 60 * 60 * 24], // days
  ['H', 1000 * 60 * 60], // hours
  ['m', 1000 * 60], // minutes
  ['s', 1000], // seconds
  ['S', 1], // million seconds
];

export function getDateString(millisecond: number, format: string) {
  let leftMillisecond: number = millisecond;
  return units.reduce((current, [name, unit]) => {
    if (current.indexOf(name) !== -1) {
      const value = Math.floor(leftMillisecond / unit);
      leftMillisecond -= value * unit;
      return current.replace(new RegExp(`${name}+`, 'g'), (match: string) => {
        const len = match.length;
        return padStart(value.toString(), len, '0');
      });
    }
    return current;
  }, format);
}
