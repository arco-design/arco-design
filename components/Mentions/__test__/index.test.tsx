import React from 'react';

import { fireEvent, render, sleep } from '../../../tests/util';
import mountTest from '../../../tests/mountTest';
import componentConfigTest from '../../../tests/componentConfigTest';
import Mentions from '..';
import { ArrowDown, Enter } from '../../_util/keycode';

mountTest(Mentions);
componentConfigTest(Mentions, 'Mentions');

let wrapper: ReturnType<typeof render>;

describe('Mentions', () => {
  afterEach(() => {
    wrapper && wrapper.unmount();
    document.body.innerHTML = '';
  });

  it('Prefix works', async () => {
    wrapper = render(<Mentions options={['beijing', 'shanghai', 'guangzhou']} />);

    expect(wrapper.find('textarea')).toHaveLength(1);

    fireEvent.keyUp(wrapper.find('textarea').item(0), { key: '@', target: { value: '@' } });

    await sleep(100);

    expect(wrapper.find('.arco-select-option')).toHaveLength(3);

    fireEvent.keyUp(wrapper.find('textarea').item(0), { key: '@bei', target: { value: '@bei' } });
    await sleep(100);

    expect(wrapper.find('.arco-select-option')).toHaveLength(1);
    expect(wrapper.find('.arco-select-option .arco-select-highlight').item(0).textContent).toBe(
      'bei'
    );
  });

  it('onChange triggered by user input', () => {
    const onChange = jest.fn();

    wrapper = render(
      <Mentions options={['beijing', 'shanghai', 'guangzhou']} onChange={onChange} />
    );

    fireEvent.change(wrapper.find('textarea').item(0), { target: { value: 'hello' } });

    expect(onChange.mock.calls[0][0]).toBe('hello');
  });

  it('onChange triggered by click option', async () => {
    const onChange = jest.fn();

    wrapper = render(
      <Mentions options={['beijing', 'shanghai', 'guangzhou']} onChange={onChange} />
    );

    fireEvent.keyUp(wrapper.find('textarea').item(0), { key: '@', target: { value: '@' } });
    await sleep(10);
    fireEvent.click(wrapper.find('.arco-select-option').item(0));
    await sleep(10);

    expect(onChange.mock.calls[0][0]).toBe('@beijing');
  });

  it('Options hide when blur textarea', async () => {
    wrapper = render(<Mentions options={['beijing', 'shanghai', 'guangzhou']} />);

    fireEvent.keyUp(wrapper.find('textarea').item(0), { key: '@', target: { value: '@' } });

    await sleep(10);
    expect(wrapper.find('.arco-select-option')).toHaveLength(3);

    fireEvent.blur(wrapper.find('textarea').item(0));
    await sleep(200);
    expect(wrapper.find('.arco-select-option')).toHaveLength(0);
  });

  it('Shortcut works', async () => {
    wrapper = render(<Mentions options={['beijing', 'shanghai', 'guangzhou']} />);

    fireEvent.keyUp(wrapper.find('textarea').item(0), { key: '@', target: { value: '@' } });
    await sleep(100);

    expect(wrapper.find('.arco-select-option')).toHaveLength(3);

    fireEvent.keyDown(wrapper.find('textarea').item(0), { keyCode: ArrowDown.code });
    fireEvent.keyDown(wrapper.find('textarea').item(0), { keyCode: Enter.code });
    await sleep(100);
    expect(wrapper.find('textarea').item(0).textContent).toBe('@shanghai');
  });
});
