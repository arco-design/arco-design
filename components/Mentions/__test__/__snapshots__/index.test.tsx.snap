// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.Mentions default 1`] = `
<div
  class="arco-mentions arco-mentions-align-textarea"
>
  <textarea
    class="arco-textarea arco-mentions-textarea"
    rows="1"
  />
  <div
    class="arco-mentions-measure"
  >
    <span
      class="arco-mentions-measure-trigger"
    />
  </div>
</div>
`;

exports[`ConfigProvider componentConfig.Mentions set className globally 1`] = `
<div
  class="arco-mentions arco-mentions-align-textarea global-default-class"
>
  <textarea
    class="arco-textarea arco-mentions-textarea"
    rows="1"
  />
  <div
    class="arco-mentions-measure"
  >
    <span
      class="arco-mentions-measure-trigger"
    />
  </div>
</div>
`;

exports[`ConfigProvider componentConfig.Mentions set className globally and component set className itself. 1`] = `
<div
  class="arco-mentions arco-mentions-align-textarea component-class"
>
  <textarea
    class="arco-textarea arco-mentions-textarea"
    rows="1"
  />
  <div
    class="arco-mentions-measure"
  >
    <span
      class="arco-mentions-measure-trigger"
    />
  </div>
</div>
`;

exports[`ConfigProvider componentConfig.Mentions set data-* property 1`] = `
<div
  class="arco-mentions arco-mentions-align-textarea"
>
  <textarea
    class="arco-textarea arco-mentions-textarea"
    data-test-name="Mentions"
    rows="1"
  />
  <div
    class="arco-mentions-measure"
  >
    <span
      class="arco-mentions-measure-trigger"
    />
  </div>
</div>
`;
